numpy/f2py/cfuncs.py

```diff
@@ -621,71 +630,127 @@
 }"""
 needs['try_pyarr_from_string'] = ['STRINGCOPYN', 'PRINTPYOBJERR', 'string']
 cfuncs['try_pyarr_from_string'] = """\
-static int try_pyarr_from_string(PyObject *obj,const string str) {
+/*
+  try_pyarr_from_string copies str[:len(obj)] to the data of an `ndarray`.
+
+  if the specified len==-1, str must be null-terminated.
+*/
+static int try_pyarr_from_string(PyObject *obj,
+                                 const string str, const int len) {
+#ifdef DEBUGCFUNCS
+fprintf(stderr,\"try_pyarr_from_string(str='%s',
+               len=%d, obj=%p)\\n\",(char*)str,len, obj);
+#endif
     PyArrayObject *arr = NULL;
-    if (PyArray_Check(obj) && (!((arr = (PyArrayObject *)obj) == NULL)))
-        { STRINGCOPYN(PyArray_DATA(arr),str,PyArray_NBYTES(arr)); }
-    return 1;
+    if (PyArray_Check(obj) && (!((arr = (PyArrayObject *)obj) == NULL))) {
+        string buf = PyArray_DATA(arr);
+        npy_intp n = len;
+        if (n == -1) {
+            /* Assuming null-terminated str. */
+            n = strlen(str);
+        }
+        if (n > PyArray_NBYTES(arr)) {
+            n = PyArray_NBYTES(arr);
+            if (n == 1 && buf[0] == '\0') {
+                /* handle the numpy.array(b'') case */
+                n = 0;
+            }
+        }
+        STRINGCOPYN(buf, str, n);
+        return 1;
+    }
 capi_fail:
     PRINTPYOBJERR(obj);
-    PyErr_SetString(#modulename#_error,\"try_pyarr_from_string failed\");
+    PyErr_SetString(#modulename#_error, \"try_pyarr_from_string failed\");
     return 0;
 }
 """
 needs['string_from_pyobj'] = ['string', 'STRINGMALLOC', 'STRINGCOPYN']
 cfuncs['string_from_pyobj'] = """\
+/*
+  Create a new string buffer from a Python string-like object.
+
+  The string buffer has given size (len) or the size of inistr when len==-1.
+
+  The string buffer may or may not be null-terminated.
```

Comment:
Doesn't the implementation ensure that it is null-terminated?

Reply:
No.
For instance, the buffers of inputs `np.array(b'1234')` or `b'1234'` are not null-terminated but the buffer of `inistr` (that is C string) is null-terminated.

Reply:
I think this needs some clarification then as there are two string buffers here - the one in `obj` and the one in `str`. It sounds like I thought this was talking about the latter, but was actually talking about the former?

Reply:
I am wrong above (didn't check STRINGMALLOC definition), see below.